src/provider/shplonk.rs

```diff
@@ -0,0 +1,660 @@
+    a: E::Fr,
+  ) -> UniPoly<E::Fr> {
+    let mut tmp = Q_x.clone();
+    tmp *= &D.evaluate(&a);
+    tmp[0] += &R_x.evaluate(&a);
+    tmp = UVKZGPoly::new(
+      tmp
+        .coeffs
+        .into_iter()
+        .map(|coeff| -coeff)
+        .collect::<Vec<E::Fr>>(),
+    );
+    let mut K_x = batched_Pi.clone();
+    K_x += &tmp;
+    K_x
+  }
+}
+
+impl<E, NE> EvaluationEngineTrait<NE> for EvaluationEngine<E, NE>
+where
+  E: MultiMillerLoop,
+  NE: NovaEngine<GE = E::G1, Scalar = E::Fr, CE = KZGCommitmentEngine<E>>,
+  E::Fr: Serialize + DeserializeOwned,
+  E::G1Affine: Serialize + DeserializeOwned,
+  E::G2Affine: Serialize + DeserializeOwned,
+  E::G1: DlogGroup<ScalarExt = E::Fr, AffineExt = E::G1Affine>,
+  <E::G1 as Group>::Base: TranscriptReprTrait<E::G1>, // Note: due to the move of the bound TranscriptReprTrait<G> on G::Base from Group to Engine
+  E::Fr: PrimeFieldBits, // TODO due to use of gen_srs_for_testing, make optional
+  E::Fr: TranscriptReprTrait<E::G1>,
+  E::G1Affine: TranscriptReprTrait<E::G1>,
+{
+  type ProverKey = KZGProverKey<E>;
+  type VerifierKey = KZGVerifierKey<E>;
+  type EvaluationArgument = EvaluationArgument<E>;
+
+  fn setup(ck: Arc<UniversalKZGParam<E>>) -> (KZGProverKey<E>, KZGVerifierKey<E>) {
+    let len = ck.length() - 1;
+    trim(ck, len)
+  }
+
+  fn prove(
+    ck: &UniversalKZGParam<E>,
+    _pk: &KZGProverKey<E>,
+    transcript: &mut <NE as NovaEngine>::TE,
+    C: &Commitment<NE>,
+    hat_P: &[E::Fr],
+    point: &[E::Fr],
+    eval: &E::Fr,
+  ) -> Result<EvaluationArgument<E>, NovaError> {
+    let x: Vec<E::Fr> = point.to_vec();
+    let ell = x.len();
+    let n = hat_P.len();
+    assert_eq!(n, 1 << ell);
+
+    // Phase 1 (similar to hyperkzg)
+    let polys = Self::compute_pi_polynomials(hat_P, point, eval);
+    let comms = Self::compute_commitments(ck, C, &polys);
+
+    // Phase 2 (similar to hyperkzg)
+    let r = HyperKZG::<E, NE>::compute_challenge(&comms, transcript);
+    let u = vec![r, -r, r * r];
+
+    let evals = Self::compute_evals(&polys, &u);
+
+    // Phase 3
+    // Compute B(x) = f_0(x) + q * f_1(x) + ... + q^(k-1) * f_{k-1}(x)
+    let q = HyperKZG::<E, NE>::get_batch_challenge(&evals, transcript);
+    let batched_Pi: UniPoly<E::Fr> = polys.into_iter().map(UniPoly::new).rlc(&q);
+
+    // Q(x), R(x) = P(x) / D(x), where D(x) = (x - r) * (x + r) * (x - r^2) = 1 * x^3 - r^2 * x^2 - r^2 * x + r^4
+    let D = UVKZGPoly::new(vec![u[2] * u[2], -u[2], -u[2], E::Fr::from(1)]);
+    let (Q_x, R_x) = batched_Pi.divide_with_q_and_r(&D).unwrap();
+
+    let C_Q = <NE::CE as CommitmentEngineTrait<NE>>::commit(ck, &Q_x.coeffs)
+      .comm
+      .to_affine();
+
+    let a = Self::compute_a(&C_Q, transcript);
+
+    // K(x) = P(x) - Q(x) * D(a) - R(a), note that R(a) should be subtracted from a free term of polynomial
+    let K_x = Self::compute_k_polynomial(&batched_Pi, &Q_x, &D, &R_x, a);
+
+    // TODO: since this is a usual KZG10 we should use it as utility instead
```

Comment:
Note the very important definition:
https://github.com/lurk-lab/arecibo/blob/7a5d7bf0e9be80b8b5c4d47bb99c67c2b058b35e/src/provider/non_hiding_kzg.rs#L224-L225
IOW, you're invoking this as a `UVKZGPoly`, but really, you can use `UniPoly` everywhere instead. I should perhaps not have made this type alias public.

Reply:
I assume this is true for Zeromorph as well, right?

Reply:
Yes